core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -437,6 +437,16 @@ public PlanOptimizers(
                                 .addAll(new CanonicalizeExpressions(metadata, typeAnalyzer).rules())
                                 .add(new OptimizeRowPattern())
                                 .build()),
+                new IterativeOptimizer(
+                        metadata,
+                        ruleStats,
+                        statsCalculator,
+                        estimatedExchangesCostCalculator,
+                        ImmutableSet.<Rule<?>>builder()
+                                .addAll(columnPruningRules)
+                                .addAll(projectionPushdownRules)
+                                .add(new PushAggregationIntoTableScan(metadata))
```

Comment:
let's push more than that

Reply:
I tried adding all of the other pushdown optimizers here, but there were done changes to other types of queries, resulting in test failures: https://github.com/trinodb/trino/runs/3125721083

Reply:
I looked at `TestAddExchangesPlans.testForcePartitioningMarkDistinctInput` failure and it was "good" (ie test wants an update). 

`TestTableScanRedirectionWithPushdown.testRedirectionBeforeDeferencePushdown` seems "bad". The name suggests early Dereference Pushdown is not expected.
Maybe we need to exclude dereference pushdown from those early ones?

`TestLogicalPlanner.testCorrelatedJoinWithLimit` -- it failed with NPE, because `PushPredicateIntoTableScan` was written in a way that it cannot be run before queries are de-correlated.
Actually, before my change https://github.com/trinodb/trino/blob/6528d39568672d0334141a77906cd84aebe01cd9/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushPredicateIntoTableScan.java#L183, it would not NPE, but it would produce incorrect results. (hence https://github.com/trinodb/trino/commit/4c75ef08e414ec1c6937aac187ca460ee903e4d1)


OTOH, adding Aggregation Pushdown alone will likely not be every effective. Aggregation pushdown may be prevented by Projection or Filter that would be pushed down, but weren't yet.


Seems like we need a bit more reordering in plan optimizers. 
Could we defer distinct aggregation rewriting optimizers after decorrelation happens, so that we can push most things into the connectors, before distinct aggregation rewriting optimizers are run?